File: leetcode/p00202/p00202.py
class Solution:
    # Time Complexity: O(logn)
    # Space Complexity: O(1)
    def isHappy(self, n: int) -> bool:
        slow = n
        fast = n
        while fast != 1 and getNext(fast) != 1:
            slow = getNext(slow)
            fast = getNext(getNext(fast))

            if slow == fast:
                return False
        
        return True
    
def getNext(n: int) -> int:
    memo = {
        0:0,
        1:1,
        2:4,
        3:9,
        4:16,
        5:25,
        6:36,
        7:49,
        8:64,
        9:81,
    }

    sum = 0
    while n != 0:
        n, remainder = divmod(n, 10)
        sum += memo[remainder]
    
    return sum

if __name__ == '__main__':
    s = Solution()

    assert s.isHappy(19) == True
    assert s.isHappy(2) == False
    assert s.isHappy(10) == True

File: leetcode/p00202/p00202.go
package p00202

// Time Complexity: O(logn)
// Space Complexity: O(logn)
func isHappy(n int) bool {
	seen := map[int]bool{}
	for {
		n = getNext(n)
		if n == 1 {
			return true
		}
		if _, exists := seen[n]; exists {
			return false
		} else {
			seen[n] = true
		}
	}
}

// Time Complexity: O(logn)
// Space Complexity: O(1)
func isHappy_twoPointers(n int) bool {
	slow := n
	fast := n
	for fast != 1 && getNext(fast) != 1 {
		slow = getNext(slow)
		fast = getNext(getNext(fast))

		if slow == fast {
			return false
		}
	}

	return true
}

// Time Complexity: O(logn)
// Space Complexity: O(1)
func isHappy_math(n int) bool {
	for {
		if n == 1 {
			return true
		} else if n == 4 {
			return false
		}

		n = getNext(n)
	}
}

// Time Complexity: O(logn)
// Space Complexity: O(1)
func getNext(n int) int {
	table := map[int]int{
		0: 0,
		1: 1,
		2: 4,
		3: 9,
		4: 16,
		5: 25,
		6: 36,
		7: 49,
		8: 64,
		9: 81,
	}

	sum := 0
	for n != 0 {
		remainder := n % 10
		sum += table[remainder]
		n /= 10
	}

	return sum
}
